### PR TITLE
Implement `Hash` for MapObserver

### PR DESCRIPTION
**Pull request by edwin1729**
* Rename the hash utility function (in MapObserver) to hash_easy

* Use hash_slice as a helper function to impl Hash trait

### CODE REVIEW

**Comment by edwin1729** [comment]
I've also tried implement `Hash` for `libafl_bolts::ownedref::OwnedMutSlice` to delete `hash_slice`. However `OwnedMapObserver::map` is `Vec<T>` not `OwnedMutSlice<T>`, which prevents removal of `hash_slice` helper in a clean way.

**Comment by domenukk** [comment]
@addisoncrump want to take a look?

**Comment by domenukk** [comment]
- I'm not a big fan of `hash_easy`, maybe rename to `_simple` to fit the rest of the names in the codebase. Also, do we really still need it?

- there's a bunch of things that don't implement `Hash` yet, such as this guy:
```rust
error[E0277]: the trait bound `observers::map::pybind::PythonMapObserverI8: core::hash::Hash` is not satisfied
    --> libafl/src/observers/map.rs:2720:34
     |
2501 | /     macro_rules! define_python_map_observer {
2502 | |         ($struct_name1:ident, $py_name1:tt, $struct_name2:ident, $py_name2:tt, $struct_name_trait:ident, $py_name_trait:tt, $datatype:ty,...
2503 | |
2504 | |             #[pyclass(unsendable, name = $py_name1)]
...    |
2720 | |             impl MapObserver for $struct_name_trait {
     | |                                  ^^^^^^^^^^^^^^^^^^ the trait `core::hash::Hash` is not implemented for `observers::map::pybind::PythonMapObserverI8`
...    |
2794 | |         };
2795 | |     }
     | |_____- in this expansion of `define_python_map_observer!`
2796 |
2797 | /     define_python_map_observer!(
2798 | |         PythonStdMapObserverI8,
2799 | |         "StdMapObserverI8",
2800 | |         PythonOwnedMapObserverI8,
...    |
2805 | |         PythonMapObserverWrapperI8
2806 | |     );
     | |_____- in this macro invocation
     |
note: required by a bound in `observers::map::MapObserver`
    --> libafl/src/observers/map.rs:85:83
     |
85   | pub trait MapObserver: HasLen + Named + Serialize + serde::de::DeserializeOwned + Hash
     |                                                                                   ^^^^ required by this bound in `MapObserver`
```

File: libafl/src/observers/map.rs [hunk 1]
```diff
@@ -70,8 +70,7 @@ fn init_count_class_16() {
 }
 
 /// Compute the hash of a slice
-fn hash_slice<T>(slice: &[T]) -> u64 {
-    let mut hasher = RandomState::with_seeds(0, 0, 0, 0).build_hasher();
+fn hash_slice<T, H: Hasher>(slice: &[T], hasher: &mut H) -> u64 {
```
Comment (addisoncrump) [inline libafl/src/observers/map.rs:73]: This entire function may replaced with `.as_slice().hash(&mut hasher)`, and should be deleted.

Comment (edwin1729) [reply libafl/src/observers/map.rs:73]: `hash_slice` should be renamed to `hash_helper`. Its used to define the `hash`  function. I can't see how to remove it.

**Comment by edwin1729** [comment]
Isn't there a way to trigger the CI pipeline? I can't get `clippy` to work, because the of the `sphinx` dependency?

**Comment by tokatoka** [comment]
if you make this PR ready for review. then CI will run. else it won't

**Comment by domenukk** [comment]
There are still some Errors in CI such as
```rust
error[E0407]: method `hash` is not a member of trait `MapObserver`
   --> libafl_targets/src/sancov_8bit.rs:199:9
    |
199 | /         fn hash(&self) -> u64 {
200 | |             let mut hasher = RandomState::with_seeds(0, 0, 0, 0).build_hasher();
201 | |             for map in unsafe { &*addr_of!(COUNTERS_MAPS) } {
202 | |                 let slice = map.as_slice();
...   |
209 | |             hasher.finish()
210 | |         }
    | |_________^ not a member of trait `MapObserver`

```

File: libafl/src/observers/map.rs (hunk 1 above)
Comment (domenukk) [reply libafl/src/observers/map.rs]: @addisoncrump any last words on this? Otherwise we'll merge

**Comment by edwin1729** [comment]
Hi, I was wondering if the commit messages are alright. I thought maybe a rebase might be necessary

Sent from Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

File: libafl_targets/src/sancov_8bit.rs [hunk 2]
```diff
@@ -208,17 +221,9 @@ mod observers {
...
-                let ptr = slice.as_ptr();
-                let map_size = slice.len() / core::mem::size_of::<u8>();
-                unsafe {
-                    hasher.write(from_raw_parts(ptr, map_size));
-                }
-            }
-            hasher.finish()
+        #[inline]
+        fn hash_simple(&self) -> u64 {
+            RandomState::with_seeds(0, 0, 0, 0).hash_one(self)
```
Comment (domenukk) [inline libafl_targets/src/sancov_8bit.rs:226]: We _could_ use the hash_std method here that can use xxh3 (but probably we don't need to) https://github.com/AFLplusplus/LibAFL/blob/8bce605503af86692e00f5d0ca9d2e72b6efa919/libafl_bolts/src/lib.rs#L283

File: libafl/src/observers/map.rs (hunk 1 above)
Comment (addisoncrump) [reply libafl/src/observers/map.rs]: It can be removed. One sec.

File: libafl_targets/src/sancov_8bit.rs (hunk 2 above)
Comment (addisoncrump) [reply libafl_targets/src/sancov_8bit.rs:226]: Why does this method even exist? We should just have a `type` definition for StdHash or something, probably.

**Comment by addisoncrump** [comment]
@domenukk I think this is good to go now. Please review latest changes and then merge it up when you're satisfied with it.

File: libafl_targets/src/sancov_8bit.rs (hunk 2 above)
Comment (domenukk) [reply libafl_targets/src/sancov_8bit.rs:226]: Not even sure if it implements any rust traits tbh

**Comment by domenukk** [comment]
Happy if you are. Thanks y'all :)

**Comment by addisoncrump** [comment]
I'll merge this when CI finishes, then :slightly_smiling_face:

File: libafl_targets/src/sancov_8bit.rs (hunk 2 above)
Comment (edwin1729) [reply libafl_targets/src/sancov_8bit.rs:226]: `hash_simple` does not implement any rust traits. If hash_std is okay then can use that instead. I don't really like it but `Hash` trait in rust doesn't seem to provide any default hashing function, so I made this

Comment (addisoncrump) [reply libafl_targets/src/sancov_8bit.rs:226]: Generally, we don't actually _want_ a default hashing function. This should be specified by the user, which is why we wanted the `Hash` impl (as this allows us to use a generic `Hasher`).

**Pull request merged**